File: sdk/csharp/Udb.Client/UdbEntityHandle.cs
using System.Collections;
using System.Globalization;
using System.Text.Json;
using Google.Protobuf;
using Google.Protobuf.WellKnownTypes;
using Udb.Client.Generated;
using Udb.Entity.V1;

namespace Udb.Client;

/// <summary>Bound relational entity/table facade over the canonical DataBroker RPCs.</summary>
public sealed class UdbEntityHandle
{
    private readonly Func<UdbMetadata> _metadata;
    private readonly Func<SelectRequest, CancellationToken, Task<RecordSet>> _select;
    private readonly Func<UpsertRequest, CancellationToken, Task<MutationResponse>> _upsert;
    private readonly Func<DeleteRequest, CancellationToken, Task<MutationResponse>> _delete;

    internal UdbEntityHandle(
        string messageType,
        IEnumerable<string> key,
        Func<UdbMetadata> metadata,
        Func<SelectRequest, CancellationToken, Task<RecordSet>> select,
        Func<UpsertRequest, CancellationToken, Task<MutationResponse>> upsert,
        Func<DeleteRequest, CancellationToken, Task<MutationResponse>> delete)
    {
        if (string.IsNullOrWhiteSpace(messageType))
        {
            throw new ArgumentException("udb: messageType is required", nameof(messageType));
        }
        MessageType = messageType;
        Key = key.ToArray();
        _metadata = metadata ?? throw new ArgumentNullException(nameof(metadata));
        _select = select ?? throw new ArgumentNullException(nameof(select));
        _upsert = upsert ?? throw new ArgumentNullException(nameof(upsert));
        _delete = delete ?? throw new ArgumentNullException(nameof(delete));
    }

    public string MessageType { get; }

    public IReadOnlyList<string> Key { get; }

    public async Task<IReadOnlyList<IReadOnlyDictionary<string, object?>>> SelectAsync(
        IReadOnlyDictionary<string, object?>? where = null,
        int limit = 0,
        CancellationToken ct = default)
    {
        var raw = await SelectRawAsync(where, limit, ct).ConfigureAwait(false);
        return DecodeRecords(raw);
    }

    public Task<RecordSet> SelectRawAsync(
        IReadOnlyDictionary<string, object?>? where = null,
        int limit = 0,
        CancellationToken ct = default)
    {
        var request = new SelectRequest
        {
            Context = Context(_metadata()),
            MessageType = MessageType,
            Filter = ToStruct(where),
            Limit = Math.Max(0, limit),
        };
        return _select(request, ct);
    }

    public Task<MutationResponse> UpsertAsync(
        IReadOnlyDictionary<string, object?> record,
        bool returnRecord = false,
        string idempotencyKey = "",
        CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(record);
        var request = new UpsertRequest
        {
            Context = Context(_metadata()),
            MessageType = MessageType,
            RecordJson = ByteString.CopyFromUtf8(RecordJson(record)),
            Payload = ToStruct(record),
            ReturnRecord = returnRecord,
            IdempotencyKey = idempotencyKey ?? string.Empty,
        };
        request.ConflictFields.AddRange(Key);
        return _upsert(request, ct);
    }

    public Task<MutationResponse> DeleteAsync(
        IReadOnlyDictionary<string, object?>? where = null,
        string idempotencyKey = "",
        CancellationToken ct = default)
    {
        var request = new DeleteRequest
        {
            Context = Context(_metadata()),
            MessageType = MessageType,
            Filter = ToStruct(where),
            IdempotencyKey = idempotencyKey ?? string.Empty,
        };
        return _delete(request, ct);
    }

    internal static UdbEntityHandle For(
        string messageType,
        IEnumerable<string>? key,
        Func<UdbMetadata> metadata,
        Func<SelectRequest, CancellationToken, Task<RecordSet>> select,
        Func<UpsertRequest, CancellationToken, Task<MutationResponse>> upsert,
        Func<DeleteRequest, CancellationToken, Task<MutationResponse>> delete)
    {
        var resolvedKey = key is null || !key.Any() ? ResolveEntityKey(messageType) : key.ToArray();
        return new UdbEntityHandle(messageType, resolvedKey, metadata, select, upsert, delete);
    }

    internal static UdbEntityHandle ForTable(
        string name,
        IEnumerable<string>? key,
        Func<UdbMetadata> metadata,
        Func<SelectRequest, CancellationToken, Task<RecordSet>> select,
        Func<UpsertRequest, CancellationToken, Task<MutationResponse>> upsert,
        Func<DeleteRequest, CancellationToken, Task<MutationResponse>> delete)
    {
        var binding = UdbIr.Entities.Values.FirstOrDefault(item => item.Table == name);
        var messageType = binding?.MessageType ?? name;
        var resolvedKey = key is null || !key.Any() ? binding?.PrimaryKeys ?? Array.Empty<string>() : key.ToArray();
        return new UdbEntityHandle(messageType, resolvedKey, metadata, select, upsert, delete);
    }

    private static IReadOnlyList<string> ResolveEntityKey(string messageType)
        => UdbIr.Entities.TryGetValue(messageType, out var binding)
            ? binding.PrimaryKeys
            : Array.Empty<string>();

    private static RequestContext Context(UdbMetadata metadata)
    {
        var context = new RequestContext
        {
            TenantId = metadata.TenantId,
            UserId = metadata.UserId,
            CorrelationId = metadata.CorrelationId,
            Purpose = metadata.Purpose,
            ServiceIdentity = metadata.ServiceIdentity,
            ProjectId = metadata.ProjectId,
            ClientCatalogVersion = metadata.ClientCatalogVersion,
            Consistency = metadata.Consistency,
            PrimaryRead = metadata.PrimaryRead,
            MaxReplicaLagMs = (ulong)Math.Max(0, metadata.MaxReplicaLagMs),
            EventualConsistencyAllowed = metadata.EventualConsistencyAllowed,
            ReadFenceJson = metadata.ReadFenceJson,
        };
        context.Scopes.AddRange(metadata.Scopes ?? Array.Empty<string>());
        return context;
    }

    private static Struct ToStruct(IReadOnlyDictionary<string, object?>? values)
    {
        var output = new Struct();
        if (values is null)
        {
            return output;
        }
        foreach (var (key, value) in values)
        {
            output.Fields.Add(key, ToValue(value));
        }
        return output;
    }

    private static Value ToValue(object? value)
    {
        if (value is null)
        {
            return new Value { NullValue = NullValue.NullValue };
        }
        if (value is JsonElement json)
        {
            return ToValue(FromJson(json));
        }
        if (value is bool boolean)
        {
            return new Value { BoolValue = boolean };
        }
        if (IsNumber(value))
        {
            var number = Convert.ToDouble(value, CultureInfo.InvariantCulture);
            if (!double.IsFinite(number))
            {
                throw new ArgumentException("udb: protobuf Struct does not support non-finite numbers");
            }
            return new Value { NumberValue = number };
        }
        if (value is IReadOnlyDictionary<string, object?> map)
        {
            return new Value { StructValue = ToStruct(map) };
        }
        if (value is IDictionary dictionary)
        {
            var nested = new Dictionary<string, object?>(StringComparer.Ordinal);
            foreach (DictionaryEntry entry in dictionary)
            {
                nested[Convert.ToString(entry.Key, CultureInfo.InvariantCulture) ?? string.Empty] = entry.Value;
            }
            return new Value { StructValue = ToStruct(nested) };
        }
        if (value is IEnumerable items and not string)
        {
            var list = new ListValue();
            foreach (var item in items)
            {
                list.Values.Add(ToValue(item));
            }
            return new Value { ListValue = list };
        }
        return new Value { StringValue = Convert.ToString(value, CultureInfo.InvariantCulture) ?? string.Empty };
    }

    private static bool IsNumber(object value)
        => value is byte or sbyte or short or ushort or int or uint or long or ulong or float or double or decimal;

    private static string RecordJson(IReadOnlyDictionary<string, object?> record)
    {
        var sorted = new SortedDictionary<string, object?>(StringComparer.Ordinal);
        foreach (var (key, value) in record)
        {
            sorted[key] = value;
        }
        return JsonSerializer.Serialize(sorted);
    }

    private static IReadOnlyList<IReadOnlyDictionary<string, object?>> DecodeRecords(RecordSet recordSet)
    {
        var rows = new List<IReadOnlyDictionary<string, object?>>(recordSet.RecordsJson.Count);
        foreach (var recordJson in recordSet.RecordsJson)
        {
            using var document = JsonDocument.Parse(recordJson.ToStringUtf8());
            var value = FromJson(document.RootElement);
            if (value is IReadOnlyDictionary<string, object?> map)
            {
                rows.Add(map);
            }
            else
            {
                throw new InvalidOperationException("udb: DataBroker returned non-object record_json");
            }
        }
        return rows;
    }

    private static object? FromJson(JsonElement element)
    {
        return element.ValueKind switch
        {
            JsonValueKind.Object => element.EnumerateObject()
                .ToDictionary(prop => prop.Name, prop => FromJson(prop.Value), StringComparer.Ordinal),
            JsonValueKind.Array => element.EnumerateArray().Select(FromJson).ToArray(),
            JsonValueKind.String => element.GetString(),
            JsonValueKind.Number => element.TryGetInt64(out var integer) ? integer : element.GetDouble(),
            JsonValueKind.True => true,
            JsonValueKind.False => false,
            JsonValueKind.Null => null,
            _ => null,
        };
    }
}

File: sdk/java/src/main/java/dev/udb/client/UdbEntityHandle.java
package dev.udb.client;

import com.google.protobuf.ByteString;
import com.google.protobuf.ListValue;
import com.google.protobuf.NullValue;
import com.google.protobuf.Struct;
import com.google.protobuf.Value;
import com.udb.entity.v1.DeleteRequest;
import com.udb.entity.v1.MutationResponse;
import com.udb.entity.v1.RecordSet;
import com.udb.entity.v1.RequestContext;
import com.udb.entity.v1.SelectRequest;
import com.udb.entity.v1.UpsertRequest;
import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.TreeMap;

/** Bound relational entity/table facade over the canonical DataBroker RPCs. */
public final class UdbEntityHandle {
  private final UdbClient client;
  private final String messageType;
  private final List<String> key;

  UdbEntityHandle(UdbClient client, String messageType, List<String> key) {
    this.client = Objects.requireNonNull(client, "client");
    if (messageType == null || messageType.isBlank()) {
      throw new IllegalArgumentException("udb: messageType is required");
    }
    this.messageType = messageType;
    this.key = List.copyOf(key == null ? List.of() : key);
  }

  public String messageType() {
    return messageType;
  }

  public List<String> key() {
    return key;
  }

  public RecordSet select(Map<String, ?> where) {
    return select(where, 0);
  }

  public RecordSet select(Map<String, ?> where, int limit) {
    SelectRequest.Builder request =
        SelectRequest.newBuilder()
            .setContext(context(client.metadata()))
            .setMessageType(messageType)
            .setFilter(toStruct(where));
    if (limit > 0) {
      request.setLimit(limit);
    }
    return client.select(request.build());
  }

  public List<String> selectJson(Map<String, ?> where) {
    RecordSet rows = select(where);
    List<String> out = new ArrayList<>(rows.getRecordsJsonCount());
    for (ByteString json : rows.getRecordsJsonList()) {
      out.add(json.toStringUtf8());
    }
    return Collections.unmodifiableList(out);
  }

  public MutationResponse upsert(Map<String, ?> record) {
    return upsert(record, false, "");
  }

  public MutationResponse upsert(Map<String, ?> record, boolean returnRecord) {
    return upsert(record, returnRecord, "");
  }

  public MutationResponse upsert(Map<String, ?> record, boolean returnRecord, String idempotencyKey) {
    Objects.requireNonNull(record, "record");
    UpsertRequest.Builder request =
        UpsertRequest.newBuilder()
            .setContext(context(client.metadata()))
            .setMessageType(messageType)
            .setRecordJson(ByteString.copyFromUtf8(toJson(record)))
            .setPayload(toStruct(record))
            .addAllConflictFields(key)
            .setReturnRecord(returnRecord);
    if (idempotencyKey != null && !idempotencyKey.isBlank()) {
      request.setIdempotencyKey(idempotencyKey);
    }
    return client.upsert(request.build());
  }

  public MutationResponse delete(Map<String, ?> where) {
    return delete(where, "");
  }

  public MutationResponse delete(Map<String, ?> where, String idempotencyKey) {
    DeleteRequest.Builder request =
        DeleteRequest.newBuilder()
            .setContext(context(client.metadata()))
            .setMessageType(messageType)
            .setFilter(toStruct(where));
    if (idempotencyKey != null && !idempotencyKey.isBlank()) {
      request.setIdempotencyKey(idempotencyKey);
    }
    return client.delete(request.build());
  }

  private static RequestContext context(UdbMetadata meta) {
    return RequestContext.newBuilder()
        .setTenantId(meta.tenantId())
        .setUserId(meta.userId())
        .setCorrelationId(meta.correlationId())
        .setPurpose(meta.purpose())
        .addAllScopes(meta.scopes())
        .setServiceIdentity(meta.serviceIdentity())
        .setProjectId(meta.projectId())
        .setClientCatalogVersion(meta.clientCatalogVersion())
        .setConsistency(meta.consistency())
        .setPrimaryRead(meta.primaryRead())
        .setMaxReplicaLagMs(meta.maxReplicaLagMs())
        .setEventualConsistencyAllowed(meta.eventualConsistencyAllowed())
        .setReadFenceJson(meta.readFenceJson())
        .build();
  }

  private static Struct toStruct(Map<String, ?> values) {
    Struct.Builder builder = Struct.newBuilder();
    if (values == null) {
      return builder.build();
    }
    for (Map.Entry<String, ?> entry : values.entrySet()) {
      builder.putFields(entry.getKey(), toValue(entry.getValue()));
    }
    return builder.build();
  }

  private static Value toValue(Object value) {
    if (value == null) {
      return Value.newBuilder().setNullValue(NullValue.NULL_VALUE).build();
    }
    if (value instanceof Boolean bool) {
      return Value.newBuilder().setBoolValue(bool).build();
    }
    if (value instanceof Number number) {
      double numeric = number.doubleValue();
      if (!Double.isFinite(numeric)) {
        throw new IllegalArgumentException("udb: protobuf Struct does not support non-finite numbers");
      }
      return Value.newBuilder().setNumberValue(numeric).build();
    }
    if (value instanceof Map<?, ?> map) {
      Struct.Builder nested = Struct.newBuilder();
      for (Map.Entry<?, ?> entry : map.entrySet()) {
        nested.putFields(String.valueOf(entry.getKey()), toValue(entry.getValue()));
      }
      return Value.newBuilder().setStructValue(nested).build();
    }
    if (value instanceof Iterable<?> items) {
      ListValue.Builder list = ListValue.newBuilder();
      for (Object item : items) {
        list.addValues(toValue(item));
      }
      return Value.newBuilder().setListValue(list).build();
    }
    if (value.getClass().isArray()) {
      ListValue.Builder list = ListValue.newBuilder();
      int len = Array.getLength(value);
      for (int idx = 0; idx < len; idx++) {
        list.addValues(toValue(Array.get(value, idx)));
      }
      return Value.newBuilder().setListValue(list).build();
    }
    return Value.newBuilder().setStringValue(String.valueOf(value)).build();
  }

  private static String toJson(Object value) {
    StringBuilder out = new StringBuilder();
    writeJson(out, value);
    return out.toString();
  }

  @SuppressWarnings("unchecked")
  private static void writeJson(StringBuilder out, Object value) {
    if (value == null) {
      out.append("null");
    } else if (value instanceof Boolean || value instanceof Byte || value instanceof Short
        || value instanceof Integer || value instanceof Long) {
      out.append(value);
    } else if (value instanceof Float || value instanceof Double) {
      double numeric = ((Number) value).doubleValue();
      if (!Double.isFinite(numeric)) {
        throw new IllegalArgumentException("udb: JSON does not support non-finite numbers");
      }
      out.append(value);
    } else if (value instanceof Number) {
      out.append(value);
    } else if (value instanceof Map<?, ?> map) {
      TreeMap<String, Object> sorted = new TreeMap<>();
      for (Map.Entry<?, ?> entry : map.entrySet()) {
        sorted.put(String.valueOf(entry.getKey()), entry.getValue());
      }
      out.append('{');
      boolean first = true;
      for (Map.Entry<String, Object> entry : sorted.entrySet()) {
        if (!first) {
          out.append(',');
        }
        first = false;
        writeJsonString(out, entry.getKey());
        out.append(':');
        writeJson(out, entry.getValue());
      }
      out.append('}');
    } else if (value instanceof Iterable<?> items) {
      out.append('[');
      boolean first = true;
      for (Object item : items) {
        if (!first) {
          out.append(',');
        }
        first = false;
        writeJson(out, item);
      }
      out.append(']');
    } else if (value.getClass().isArray()) {
      out.append('[');
      int len = Array.getLength(value);
      for (int idx = 0; idx < len; idx++) {
        if (idx > 0) {
          out.append(',');
        }
        writeJson(out, Array.get(value, idx));
      }
      out.append(']');
    } else {
      writeJsonString(out, String.valueOf(value));
    }
  }

  private static void writeJsonString(StringBuilder out, String value) {
    out.append('"');
    for (int idx = 0; idx < value.length(); idx++) {
      char ch = value.charAt(idx);
      switch (ch) {
        case '"' -> out.append("\\\"");
        case '\\' -> out.append("\\\\");
        case '\b' -> out.append("\\b");
        case '\f' -> out.append("\\f");
        case '\n' -> out.append("\\n");
        case '\r' -> out.append("\\r");
        case '\t' -> out.append("\\t");
        default -> {
          if (ch < 0x20) {
            out.append(String.format("\\u%04x", (int) ch));
          } else {
            out.append(ch);
          }
        }
      }
    }
    out.append('"');
  }
}
